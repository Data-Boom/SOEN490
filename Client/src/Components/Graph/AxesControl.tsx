import * as svg from 'save-svg-as-png'

import { Box, Button, FormControl, Grid, InputLabel, MenuItem, Modal, Paper, Select, Typography, makeStyles } from "@material-ui/core"
import { IDatasetModel, IVariable } from "../../Models/Datasets/IDatasetModel"
import { IVariableAndUnitModel, IVariableUnits } from '../../Models/Datasets/IVariableModel'
import React, { useState } from "react"

import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown'
import KeyboardArrowUpIcon from '@material-ui/icons/KeyboardArrowUp'

interface IProps {
    datasets: IDatasetModel[],
    variables: string[],
    axes: IVariableAndUnitModel,
    onAxesChange: (axes: IVariableAndUnitModel) => void
}

export const AxesControl = (props: IProps) => {

    //const [completeDatasets, setCompleteDatasets] = useState<IDatasetModel[]>(props.datasets)
    //todo unhardcode the variables
    const [showSettings, setSettingsToggle] = useState(false)
    //const [axes, setAxes] = useState<IVariableAndUnitModel>(props.axes)

    const [xVariableMissing, setXVariableMissing] = useState([])
    const [yVariableMissing, setYVariableMissing] = useState([])
    const [xUnits, setXUnits] = useState([])
    const [yUnits, setYUnits] = useState([])

    const setUnitType = (variable: string, type: string): string => {
        let measurement = ''
        IVariableUnits.forEach(variables => {
            const units = variables.units
            variables.variableNames.forEach(name => {
                if (type == name) {
                    if (variable == 'x') {
                        setXUnits(units)
                        measurement = units[0]
                    }
                    else if (variable == 'y') {
                        setYUnits(units)
                        measurement = units[0]
                    }
                }
            })
        })
        return measurement
    }

    const handleSettingsClick = () => {
        setSettingsToggle(!showSettings)
    }

    const handleXVariableChange = (event: React.ChangeEvent<{ value: unknown }>) => {
        let sameVariable = false, tempVariable = '', xUnit = '', yUnit = ''
        if (props.axes.yVariableName == (event.target.value as string)) {
            tempVariable = props.axes.xVariableName
            sameVariable = true
            yUnit = setUnitType('y', tempVariable)
            checkYVariablesExist(tempVariable, props.datasets)
        }
        xUnit = setUnitType('x', event.target.value as string)
        checkXVariablesExist(event.target.value as string, props.datasets)
        if (sameVariable == true) {
            props.onAxesChange({ ...props.axes, xVariableName: event.target.value as string, yVariableName: tempVariable, xVariableUnits: xUnit, yVariableUnits: yUnit })
        }
        else {
            props.onAxesChange({ ...props.axes, xVariableName: event.target.value as string, xVariableUnits: xUnit })
        }
    }
    const handleYVariableChange = (event: React.ChangeEvent<{ value: unknown }>) => {
        let sameVariable = false, tempVariable = '', xUnit = '', yUnit = ''
        if (props.axes.xVariableName == (event.target.value as string)) {
            tempVariable = props.axes.yVariableName
            sameVariable = true
            xUnit = setUnitType('x', tempVariable)
            checkXVariablesExist(tempVariable, props.datasets)
        }
        yUnit = setUnitType('y', event.target.value as string)
        checkYVariablesExist(event.target.value as string, props.datasets)
        if (sameVariable == true) {
            props.onAxesChange({ ...props.axes, yVariableName: event.target.value as string, xVariableName: tempVariable, xVariableUnits: xUnit, yVariableUnits: yUnit })
        }
        else {
            props.onAxesChange({ ...props.axes, yVariableName: event.target.value as string, yVariableUnits: yUnit })
        }
    }
    const handleXUnitChange = (event: React.ChangeEvent<{ value: unknown }>) => {
        props.onAxesChange({ ...props.axes, xVariableUnits: event.target.value as string })
    }
    const handleYUnitChange = (event: React.ChangeEvent<{ value: unknown }>) => {
        props.onAxesChange({ ...props.axes, yVariableUnits: event.target.value as string })
    }

    const checkXVariablesExist = (type: string, datasets: IDatasetModel[]) => {
        const missingDatasets = []
        datasets.forEach(dataset => {
            let exists = false
            dataset.data.variables.forEach(variableName => {
                if (variableName.name == type) {
                    exists = true
                }
            })
            if (exists == false) {
                missingDatasets.push(dataset.dataset_name)
            }
        })
        setXVariableMissing(missingDatasets)
    }

    const checkYVariablesExist = (type: string, datasets: IDatasetModel[]) => {
        const missingDatasets = []
        console.log(type)
        datasets.forEach(dataset => {
            let exists = false
            dataset.data.variables.forEach(variableName => {
                if (variableName.name == type) {
                    exists = true
                }
            })
            if (exists == false) {
                missingDatasets.push(dataset.dataset_name)
            }
        })
        console.log(missingDatasets)
        setYVariableMissing(missingDatasets)
    }

    return (
        <>
            <Grid container direction='row'>
                <Grid item xs={12}>
                    <Button id='settingsToggle' variant="contained" onClick={handleSettingsClick} color="primary">
                        Settings
              {showSettings ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
                    </Button>
                </Grid>
                {showSettings &&
                    <>
                        <Grid item xs={4}>
                            <FormControl>
                                <InputLabel id="xVariable">X Variable</InputLabel>
                                <Select
                                    labelId="xVariable"
                                    id="xVariable"
                                    value={props.axes.xVariableName}
                                    autoWidth={true}
                                    onChange={handleXVariableChange}
                                >
                                    {props.variables.map(type => (
                                        <MenuItem value={type}>{type}</MenuItem>
                                    ))}
                                </Select>
                            </FormControl>
                        </Grid>
                        <Grid item xs={4}>
                            <FormControl>
                                <InputLabel id="xUnits">X Units</InputLabel>
                                <Select
                                    labelId="xUnits"
                                    id="xUnits"
                                    value={props.axes.xVariableUnits}
                                    autoWidth={true}
                                    onChange={handleXUnitChange}
                                >
                                    {xUnits.map(type => (
                                        <MenuItem value={type}>{type}</MenuItem>
                                    ))}
                                </Select>
                            </FormControl>
                        </Grid>
                        <Grid item xs={4}>
                            <Typography align="center">
                                Datasets Missing X: {xVariableMissing.toString()}
                            </Typography>
                        </Grid>
                        <Grid item xs={4}>
                            <FormControl>
                                <InputLabel id="yVariable">Y Variable</InputLabel>
                                <Select
                                    labelId="yVariable"
                                    id="yVariable"
                                    value={props.axes.yVariableName}
                                    autoWidth={true}
                                    onChange={handleYVariableChange}
                                >
                                    {props.variables.map(type => (
                                        <MenuItem value={type}>{type}</MenuItem>
                                    ))}
                                </Select>
                            </FormControl>
                        </Grid>
                        <Grid item xs={4}>
                            <FormControl>
                                <InputLabel id="yUnits">Y Units</InputLabel>
                                <Select
                                    labelId="yUnits"
                                    id="yUnits"
                                    value={props.axes.yVariableUnits}
                                    autoWidth={true}
                                    onChange={handleYUnitChange}
                                >
                                    {yUnits.map(type => (
                                        <MenuItem value={type}>{type}</MenuItem>
                                    ))}
                                </Select>
                            </FormControl>
                        </Grid>
                        <Grid item xs={4}>
                            <Typography align="center">
                                Datasets Missing Y: {yVariableMissing.toString()}
                            </Typography>
                        </Grid>
                    </>
                }
            </Grid>
        </>
    )
}