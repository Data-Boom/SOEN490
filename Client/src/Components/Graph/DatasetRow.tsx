import { Box, Grid, IconButton, Typography } from "@material-ui/core"

import DeleteIcon from '@material-ui/icons/Delete'
import { IDatasetModel } from "../../Models/Datasets/IDatasetModel"
import React from 'react'
import { classStyles } from "../../appTheme"

interface IProps {
  dataset: IDatasetModel,
  onRemoveDatasetClick: (datasetId: number) => void
}

export const DatasetRow = (props: IProps) => {
  return (
    <Grid item>
      <Box className={classStyles().datasetBorder} p={1}>
        <Grid item container alignItems='center' justify='space-between'>
          <Grid container justify="flex-start" xs={9}>
            <Typography variant="body2" noWrap>
              {props.dataset.dataset_name}
            </Typography>
          </Grid>
          <IconButton onClick={() => props.onRemoveDatasetClick(props.dataset.id)}>
            <DeleteIcon color='primary' />
          </IconButton>
        </Grid>
      </Box>
    </Grid>
  )
}
