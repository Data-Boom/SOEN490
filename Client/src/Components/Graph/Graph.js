import * as d3 from "d3"

import React, { useEffect, useState } from 'react'

import { Button } from "@material-ui/core"

export default function Graph(props) {

  const margin = {
    top: 50,
    right: 70,
    bottom: 50,
    left: 70
  },
    outerWidth = props && props.outerWidth || 768,
    outerHeight = props && props.outerHeight || 500,
    width = outerWidth - margin.left - margin.right,
    height = outerHeight - margin.top - margin.bottom

  const [isXLog, setXToggle] = useState(false)
  const [isYLog, setYToggle] = useState(false)

  const handleXScaleClick = () => {
    setXToggle(!isXLog)
  }

  const handleYScaleClick = () => {
    setYToggle(!isYLog)
  }

  let active1 = null, active2 = null, active3 = null, active4 = null
  let active = [null, null, null, null]

  const ref = React.useRef(null)
  const changeX = React.useRef(null)
  const changeY = React.useRef(null)

  const getScale = (isLog, rangeFrom, rangeTo) => {
    let scale
    if (isLog) {
      scale = d3.scaleLog().domain([1, 10])
    }
    else {
      scale = d3.scaleLinear().domain([0, 10])
    }
    return scale.range([rangeTo, rangeFrom]).nice()
  }

  useEffect(() => {
    //cleans up all the points from the graph
    d3.select(ref.current).selectAll("*").remove()
    //This part creates the canvas for our graph
    var svg = d3.select(ref.current) // FIX WARNING "Warning: A string ref, "canvas", has been found within a strict mode tree. String refs are a source of potential bugs and should be avoided. We recommend using useRef() or createRef() instead. Learn more about using refs safely here: https://fb.me/react-strict-mode-string-ref"
      .append("svg")
      .attr("width", outerWidth)
      .attr("height", outerHeight)
      .append("g")
      .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

    //This part creates the axis/scales used for the data.
    const xScale = getScale(isXLog, width, 0)
    const yScale = getScale(isYLog, 0, height)

    //Calls the function to create the axis
    const xAxis = svg.append("g")
      .attr("transform", "translate(0," + height + ")")
      .call(d3.axisBottom(xScale));
    const yAxis = svg.append("g")
      .call(d3.axisLeft(yScale));

    //This part creates an area where points will not be drawn if they are not within this area.
    var clip = svg.append("defs").append("SVG:clipPath")
      .attr("id", "clip")
      .append("SVG:rect")
      .attr("width", width)
      .attr("height", height)
      .attr("x", 0)
      .attr("y", 0);

    //This is the part that creates the points
    var scatter = svg.append("g")
      .attr("clip-path", "url(#clip)");

    scatter
      //the myDots is a unique name, this is to refer to these dots.
      .selectAll("myDots")
      //Datalist is the list of list of points, the enter creates a loop through each one.
      .data(props && props.datalist)
      .enter()
      .append('g')
      //This gives each list of points a different colour, verifies if its in the list and changes it if it is
      .attr("fill", function (d) {
        var x = props.datalist.indexOf(d);
        return props.colourslist[x];
      })
      .attr("id", function (d) {
        var x = props.datalist.indexOf(d);
        return "id" + props.IDList[x];
      })
      .attr("stroke", "black")
      .attr("stroke-width", 2)

      //This is for the inner list (dataset1, etc). again, the myPoints is a unique name
      .selectAll("myPoints")
      .data(function (d) {
        return d;
      })
      .enter()
      .append("circle")
      .style("opacity", 1)
      .attr("r", 10)
      //This gives the value of the x position
      .attr("cx", function (d) {
        return xScale(d["x"]);
      })
      //this gives the value of the y position
      .attr("cy", function (d) {
        return yScale(d["y"]);
      })

    //This allows the user to zoom in/out onto the graph.
    var zoom = d3.zoom()
      .scaleExtent([1, 20])
      .extent([[0, 0], [width, height]])
      .on("zoom", updateGraph);
    //This rectangle is the area in which the user can zoom into.
    svg.append("rect")
      .attr("fill", "none")
      .attr("width", width)
      .attr("height", height)
      .style("pointer-events", "all")
      .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')')
      .call(zoom)

    //Legend on the graph
    svg.selectAll("mylegendDots")
      .data(props && props.datalist)
      .enter()
      //linking an image to be part of the legend instead of a circle
      .append('image')
      .attr('xlink:href', "https://cdn2.iconfinder.com/data/icons/flat-ui-icons-24-px/24/eye-24-512.png")
      .attr('width', 20)
      .attr('height', 20)
      .attr("x", 600)
      .attr("y", function (d, i) { return 290 + i * 25 }) // 100 is where the first dot appears. 25 is the distance between dots
      .attr("r", 7)
      .attr("id", function (d) {
        var x = props.datalist.indexOf(d);
        return "legenddotid" + props.IDList[x];
      })
      //An option to make the legend hide and show datasets when clicked on.
      .on("click", function (d) {
        var x = props.datalist.indexOf(d);
        scatter
          .selectAll("#id" + props.IDList[x])
          .style("opacity", function () {
            if (active[x] == null) {
              active[x] = !d3.select(this).style('opacity') ? 1 : 0;
            }
            else {
              active[x] = !active[x] ? 1 : 0;
            }
            return active[x];
          })
        //This will change the icon when the datasets are hidden or visible
        svg
          .selectAll("#legenddotid" + props.IDList[x])
          .attr("xlink:href", function () {
            if (active[x] == 0) {
              return "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSJ9809ku1l9OC6QM7kT2UimZhtywkCrB_0aQ&usqp=CAU";
            }
            else {
              return "https://cdn2.iconfinder.com/data/icons/flat-ui-icons-24-px/24/eye-24-512.png";
            }
          })
      })

    //The name labels for the datasets mentioned in the legend
    svg.selectAll("mylabels")
      .data(props && props.datalist)
      .enter()
      .append("text")
      .attr("x", 625)
      .attr("y", function (d, i) { return 300 + i * 25 }) // 100 is where the first dot appears. 25 is the distance between dots
      .attr("fill", function (d) {
        var x = props.datalist.indexOf(d);
        return props.colourslist[x];
      })
      .text(function (d) {
        var x = props.datalist.indexOf(d);
        return "dataset" + props.IDList[x];
      })
      .attr("text-anchor", "left")
      .style("alignment-baseline", "middle")

    //delete icon to delete the data
    svg.selectAll("deleteIcon")
      .data(props && props.datalist)
      .enter()
      .append("text")
      .attr("x", 560)
      .attr("y", function (d, i) { return 305 + i * 25 }) // 100 is where the first dot appears. 25 is the distance between dots
      .attr("fill", '#69a3b2')
      .text("delete")



    //This function modifies the graph upon zooming in/out by updating the axes and points.
    function updateGraph() {
      const newXScale = d3.event.transform.rescaleX(xScale)
      xAxis.call(d3.axisBottom(newXScale))
      scatter
        .selectAll("circle")
        .attr('cx', function (d) { return (newXScale(d["x"])) })

      const newYScale = d3.event.transform.rescaleY(yScale)
      yAxis.call(d3.axisLeft(newYScale))
      scatter
        .selectAll("circle")
        .attr('cy', function (d) { return (newYScale(d["y"])) })
    }
  }, [props, isXLog, isYLog])

  function removeDatasets() {

  }

  function removeDatasets() {

  }

  return (
    <>
      <svg
        ref={ref}
        width={outerWidth}
        height={outerHeight}
        viewBox={`0 0 ${outerWidth} ${outerHeight}`}
        preserveAspectRatio="xMidYMid meet">
      </svg>
      <div>
        <Button onClick={handleXScaleClick} color="primary">Change X Scale</Button>
        <Button onClick={handleYScaleClick} color="primary">Change Y Scale</Button>
      </div>
    </>
  )
}