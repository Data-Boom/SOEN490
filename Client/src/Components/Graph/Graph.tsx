import * as am4charts from "@amcharts/amcharts4/charts"
import * as am4core from "@amcharts/amcharts4/core"

import React, { useEffect } from 'react'

import { Box } from "@material-ui/core"
import { IAxisStateModel } from "../../Models/Graph/IGraphStateModel"
import { IGraphDatasetModel } from "../../Models/Graph/IGraphDatasetModel"
import am4themes_material from "@amcharts/amcharts4/themes/animated"

interface IProps {
  datasets: IGraphDatasetModel[],
  axes: IAxisStateModel[]
}

am4core.useTheme(am4themes_material)

export default function Graph(props: IProps) {
  const { datasets, axes } = { ...props }

  useEffect(() => setUpGraph(), [datasets])

  const addSeries = (chart) => {
    datasets.forEach(dataset => {
      const lineSeries = chart.series.push(new am4charts.CandlestickSeries())
      const bullet = lineSeries.bullets.push(new am4charts.CircleBullet())
      //todo make tooltip nice
      bullet.tooltipText = `${dataset.name}
      ${axes[0].variableName}: {${dataset.id}x} ${axes[0].units}
      ${axes[1].variableName}: {${dataset.id}y} ${axes[1].units}`

      lineSeries.dataFields.valueX = `${dataset.id}x`
      lineSeries.dataFields.valueY = `${dataset.id}y`
    })
  }

  const buildDataForGraph = (datasets: IGraphDatasetModel[]): any[] => {
    const data = []
    for (let datasetIndex = 0; datasetIndex < datasets.length; datasetIndex++) {
      const dataset = datasets[datasetIndex]
      mergeDatasetPoints(dataset, data)
    }

    return data
  }

  const mergeDatasetPoints = (dataset: IGraphDatasetModel, data: any[]) => {
    if (!dataset || dataset.isHidden) {
      return
    }
    //for all current dataset points:
    for (let pointIndex = 0; pointIndex < dataset.points.length; pointIndex++) {
      //create new point that has previous points info and this one's
      const mergedPoint = { ...data[pointIndex], [`${dataset.id}x`]: dataset.points[pointIndex].x, [`${dataset.id}y`]: dataset.points[pointIndex].y }
      data[pointIndex] = mergedPoint
    }
  }

  const subscribeToChartEvents = (chart: am4charts.XYChart, xAxis: am4charts.Axis, yAxis: am4charts.Axis) => {
    chart.events.on("ready", function () {
      //todo zoom to values or indeces?
      axes[0].zoomStartIndex &&
        axes[0].zoomEndIndex &&
        xAxis.zoomToIndexes(axes[0].zoomStartIndex, axes[0].zoomEndIndex)

      axes[1].zoomStartIndex &&
        axes[1].zoomEndIndex &&
        yAxis.zoomToIndexes(axes[1].zoomStartIndex, axes[1].zoomEndIndex)
    })
  }

  const updateAxis = (xAxis, yAxis) => {
    xAxis.title.text = `${axes[0].variableName}, ${axes[0].units}`
    xAxis.logarithmic = axes[0].logarithmic || false
    xAxis.renderer.minGridDistance = 40

    yAxis.logarithmic = axes[1].logarithmic || false
    yAxis.title.text = `${axes[1].variableName}, ${axes[1].units}`
  }

  const setUpGraph = () => {
    const chart = am4core.create("chartdiv", am4charts.XYChart)
    chart.data = buildDataForGraph(datasets)

    const xAxis = chart.xAxes.push(new am4charts.ValueAxis())
    const yAxis = chart.yAxes.push(new am4charts.ValueAxis())
    updateAxis(xAxis, yAxis)
    addSeries(chart)

    chart.scrollbarX = new am4core.Scrollbar()
    chart.scrollbarY = new am4core.Scrollbar()
    chart.cursor = new am4charts.XYCursor()
    chart.exporting.menu = new am4core.ExportMenu()

    subscribeToChartEvents(chart, xAxis, yAxis)
  }

  return (
    <Box id="chartdiv" width='100%' height='750px'>
    </Box>
  )
}