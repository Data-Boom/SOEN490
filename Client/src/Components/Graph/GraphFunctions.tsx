import { IDatasetModel, IVariable } from "../../Models/Datasets/IDatasetModel"
import { IGraphDatasetModel, IGraphDatasetState, IGraphPoint, newGraphDataset, toGraphDatasetState } from "../../Models/Graph/IGraphDatasetModel"

import { IDatasetRowModel } from "../../Models/Datasets/IDatasetRowModel"
import { IGraphStateModel } from "../../Models/Graph/IGraphStateModel"

export const buildXYPoints = (dataset: IDatasetModel, xVariableName: string, yVariableName: string): IGraphPoint[] => {
  const xIndex = getVariableIndex(dataset.data.variables, xVariableName)
  const yIndex = getVariableIndex(dataset.data.variables, yVariableName)
  //if either is -1 means at least one variable is not on the dataset and cannot be graphed
  if (xIndex === -1 || yIndex === -1) {
    return []
  }
  const points: IGraphPoint[] = []
  for (let i = 0; i < dataset.data.contents.length; i++) {
    const x: number = dataset.data.contents[i].point[xIndex]
    const y: number = dataset.data.contents[i].point[yIndex]
    const point: IGraphPoint = { x: x, y: y }
    points.push(point)
  }
  return points
}

export const getGraphDatasets = (completeDatasets: IDatasetModel[], graphState: IGraphStateModel): IGraphDatasetModel[] => {
  const updatedGraphDatasets = []
  completeDatasets.forEach(completeDataset => {
    const existingGraphDatasetIndex = graphState.datasets.findIndex(dataset => dataset.id == completeDataset.id)
    if (existingGraphDatasetIndex == -1) {
      console.error(`failed to build graph datasets, index for ${completeDataset.id} was not found in the state`)
    }
    const updatedGraphDataset: IGraphDatasetModel = { ...graphState.datasets[existingGraphDatasetIndex] } as any
    updatedGraphDataset.name = completeDataset.dataset_name
    updatedGraphDataset.points = buildXYPoints(completeDataset, graphState.axes[0].variableName, graphState.axes[1].variableName)
    updatedGraphDatasets.push(updatedGraphDataset)
  })

  return updatedGraphDatasets
}

export const toDatasetRows = (datasets: IDatasetModel[], graphDatasets: IGraphDatasetState[]): IDatasetRowModel[] => {
  const datasetRows: IDatasetRowModel[] = []

  for (let i = 0; i < datasets.length; i++) {
    const datasetRow: IDatasetRowModel = { ...graphDatasets[i], id: datasets[i].id, name: datasets[i].dataset_name, isInitiallyHidden: graphDatasets[i].isHidden }
    datasetRows.push(datasetRow)
  }
  return datasetRows
}

export const getVariableIndex = (variables: IVariable[], varName: string): number => {
  return variables.findIndex(variable => variable.name === varName)
}