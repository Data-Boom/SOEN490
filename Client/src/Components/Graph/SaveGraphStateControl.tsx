import { Field, Form, Formik } from "formik"

import { Grid } from "@material-ui/core"
import { IGraphStateModel } from "../../Models/Graph/IGraphStateModel"
import React from 'react'
import Typography from "material-ui/styles/typography"

interface IProps {
  graphState: IGraphStateModel
}

export const SaveGraphStateControl = (props: IProps) => {
  const { graphState } = { ...props }

  const saveGraphState = () => {

  }

  return (
    <Formik
      initialValues={defaultSearchDatasetsModel}
      validationSchema={searchDatasetsValidationSchema}
      onSubmit={transformAndSubmit}
    >
      <Form>
        <Typography variant='h4' align="left">Search</Typography>
        <Grid container spacing={4}>
          <Grid item sm={2}>
            <Field name="firstName" label='First Name' component={MuiTextFieldFormik} />
          </Grid>
        </Grid>
      </Form>
    </Formik >
  )
}