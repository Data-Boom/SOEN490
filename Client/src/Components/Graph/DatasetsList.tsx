import { Box, Grid } from "@material-ui/core"

import { DatasetRow } from "./DatasetRow"
import { IDatasetModel } from "../../Models/Datasets/IDatasetModel"
import React from 'react'

interface IProps {
  datasets: IDatasetModel[],
  onRemoveDatasetClick: (datasetId: number) => void
}

export const DatasetsList = (props: IProps) => {

  const renderDatasetRows = () => {
    return props && props.datasets && props.datasets.map(dataset => {
      return (<DatasetRow dataset={dataset} key={dataset.id} onRemoveDatasetClick={props.onRemoveDatasetClick} />)
    })
  }

  return (
    <>
      <Grid item container direction='column' spacing={3}>
        <Box className="result-checkbox" >
          {renderDatasetRows()}
        </Box>
      </Grid>
    </>
  )
}