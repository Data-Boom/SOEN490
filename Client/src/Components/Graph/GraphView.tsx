import { Box, Grid } from "@material-ui/core"
import { IGraphStateModel, newGraphState } from "../../Models/Graph/IGraphStateModel"
import React, { useState } from "react"

import Graph from './Graph'
import { GraphStateControl } from "./GraphControls/GraphStateControl"
import { IDatasetModel } from "../../Models/Datasets/IDatasetModel"
import { IGraphDatasetModel } from '../../Models/Graph/IGraphDatasetModel'
import { getGraphDatasets } from "./GraphFunctions"
import { useParams } from "react-router"

interface IGraphViewParams {
  graphStateId: string
}

export default function GraphView() {
  const { graphStateId } = useParams<IGraphViewParams>()

  const [graphDatasets, setGraphDatasets] = useState<IGraphDatasetModel[]>([])
  const [graphState, setGraphState] = useState<IGraphStateModel>({ ...newGraphState, id: graphStateId })

  const handleGraphStateChanged = (graphState: IGraphStateModel, completeDatasets: IDatasetModel[]) => {
    const graphDatasets = getGraphDatasets(completeDatasets, graphState)
    setGraphState(graphState)
    setGraphDatasets(graphDatasets)
  }

  return (
    <>
      <Box ml={8}>
        <Grid container spacing={3}>
          <Grid item container sm={7} >
            <Graph
              datasets={graphDatasets}
              axes={graphState.axes}
            />
          </Grid>
          <Grid item sm={5}>
            <Box ml={5} mr={5} mt={5}>
              <GraphStateControl
                graphState={graphState}
                onGraphStateChange={handleGraphStateChanged}
              />
            </Box>
          </Grid>
        </Grid>
      </Box >
    </>
  )
}