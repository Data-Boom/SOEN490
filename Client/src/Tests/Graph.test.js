import React from 'react';
import renderer from "react-test-renderer";
import { shallow } from 'enzyme';
import Graph from '../Components/Graph/Graph'

describe('Graph Component', () => {
    it("should match snapshot", () => {
        const tree = renderer.create(<Graph />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    /*it("should change the xscale upon toggle", () => {

    });

    it("should change the xscale upon toggle", () => {

    });*/
});