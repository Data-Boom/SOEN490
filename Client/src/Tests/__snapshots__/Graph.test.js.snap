// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Component should match snapshot 1`] = `
Array [
  <svg
    height={500}
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 768 500"
    width={768}
  />,
  <div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Change X Scale
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Change Y Scale
      </span>
    </button>
  </div>,
]
`;
